cmd/kubeadm/app/cmd/phases/join/controlplanejoin.go

```diff
@@ -0,0 +1,256 @@
+/*
+Copyright 2019 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package phases
+
+import (
+	"fmt"
+	"path/filepath"
+
+	"github.com/pkg/errors"
+	clientset "k8s.io/client-go/kubernetes"
+	kubeadmapi "k8s.io/kubernetes/cmd/kubeadm/app/apis/kubeadm"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/options"
+	"k8s.io/kubernetes/cmd/kubeadm/app/cmd/phases/workflow"
+	cmdutil "k8s.io/kubernetes/cmd/kubeadm/app/cmd/util"
+	kubeadmconstants "k8s.io/kubernetes/cmd/kubeadm/app/constants"
+	"k8s.io/kubernetes/cmd/kubeadm/app/phases/controlplane"
+	etcdphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/etcd"
+	markcontrolplanephase "k8s.io/kubernetes/cmd/kubeadm/app/phases/markcontrolplane"
+	uploadconfigphase "k8s.io/kubernetes/cmd/kubeadm/app/phases/uploadconfig"
+	"k8s.io/kubernetes/pkg/util/normalizer"
+)
+
+var (
+	controlPlaneJoinExample = normalizer.Examples(`
+		# Generates all static Pod manifest files for control plane components,
+		# functionally equivalent to what is generated by kubeadm init.
+		kubeadm init phase control-plane all
+		# Generates all static Pod manifest files using options read from a configuration file.
+		kubeadm init phase control-plane all --config config.yaml
+		`)
+
+	controlPlanePhaseProperties = map[string]struct {
+		name  string
+		short string
+	}{
+		kubeadmconstants.KubeAPIServer: {
+			name:  "apiserver",
+			short: getPhaseDescription(kubeadmconstants.KubeAPIServer),
+		},
+		kubeadmconstants.KubeControllerManager: {
+			name:  "controller-manager",
+			short: getPhaseDescription(kubeadmconstants.KubeControllerManager),
+		},
+		kubeadmconstants.KubeScheduler: {
+			name:  "scheduler",
+			short: getPhaseDescription(kubeadmconstants.KubeScheduler),
+		},
+	}
+)
+
+type controlPlaneJoinData interface {
+	Cfg() *kubeadmapi.JoinConfiguration
+	ClientSetFromFile(string) (*clientset.Clientset, error)
+	InitCfg() (*kubeadmapi.InitConfiguration, error)
+}
+
+func getPhaseDescription(component string) string {
+	return fmt.Sprintf("Generates the %s static Pod manifest", component)
+}
+
+func getControlPlaneJoinPhaseFlags() []string {
+	return []string{
+		options.APIServerAdvertiseAddress,
+		options.APIServerBindPort,
+		options.CfgPath,
+		options.ControlPlane,
+		options.NodeName,
+		options.TokenDiscovery,
+		options.TokenDiscoveryCAHash,
+		options.TokenDiscoverySkipCAHash,
```

Comment:
@ereslibre @rosti can you please double check these.
i think we need `options.KubeconfigPath` too.

Reply:
I don't think we need `TokenDiscovery*` stuff in here, since we don't make use of it. It's used to authenticate the kebelet to the API Server we are joining to, but this subject of a different phase.
On the other hand, we may need `options.FileDiscovery` and actually take that into account when looking for the kubeconfig.

Reply:
@rosti we need them for the same reasons as https://github.com/kubernetes/kubernetes/pull/73145#discussion_r253142805

Reply:
@neolit123 added options.KubeconfigPath here